### PR TITLE
Add 3D fused-ssim

### PR DESCRIPTION
**Pull request by PaPieta**
Hello! I needed a 3D (volumetric) version of your implementation, and managed to achieve ~11x acceleration compared to the baseline pytorch-mssim:

  

* The code passes tests comparable to your own (```tests/test_3D.py```). Note the lowered ```iterations``` number, with the original 100, the baseline takes forever.
* Of course, it is still slower than your 2D code for the same number of pixels.
* **Main Limitation:** Only a CUDA version is available at the moment.

This is a branch made specifically to make the merge easy. I also made a more stand-alone version in the main branch. You can check it out [here](https://github.com/PaPieta/fused-ssim3D/tree/main).

Let me know if you need anything more.

### How to use

All existing logic is preserved. Running the 3D version can be done with:

```python
import torch
from fused_ssim import fused_ssim3d

# predicted_image, gt_image: [BS, CH, D, H, W]
# predicted_image is differentiable
gt_image = torch.rand(2, 3, 96, 96, 96)
predicted_image = torch.nn.Parameter(torch.rand_like(gt_image))
ssim_value = fused_ssim3d(predicted_image, gt_image)
```

### Implementation details

A natural extension of the 2D fused-ssim would be to load a 3D neighbourhood into shared memory (```sTile```). While on paper the fastest, this method is strongly limited by shared memory size constraints. It is possible to make it work by heavy reuse of a single large array (see e.g [this kernel version](https://github.com/PaPieta/fused-ssim3D/blob/609a2f185be620cd4bda634b377c5c75f0af9b74/ssim3d.cu)), but the shared memory is still big enough to only enable a handful (often 1-3) blocks per SM, which in turn greatly affects latency hiding during memory access. With this approach, the acceleration was ~5x.

An alternative approach (implemented here) is to:
1. Perform 2D Gaussian convolutions on each slice (similarly to the ssim architecture),
2. Save intermediate results to global memory,
3. Make each thread calculate the Z axis convolution and final SSIM along the whole Z row (depth).

The last step is achieved by loading appropriate data from global memory into a ring buffer. With each step, only a missing data point is loaded, and the "center pixel" index is adjusted.

### CODE REVIEW

**Comment by PaPieta** [comment]
Thanks! Yes, I primarily plan to use it for CT data. I don't want to share too many details just yet, I'm still trying to figure things out, but it probably won't be a surprise that it is related to Gaussian splatting.

**Comment by PaPieta** [comment]
Thanks for the merge. I just realized I forgot to mention that I didn't make any changes in the README, so it may be a good idea to do that. Especially in line:
https://github.com/rahul-goel/fused-ssim/blob/0ac916a05f48f8e477dc8ce1651a346b65ce2072/README.md?plain=1#L201C1-L201C24

I can prepare a more extensive readme update myself as well, I just don't want to interfere too much with your setup.

**Comment by rahul-goel** [comment]
Hi.

A small paragraph about 3D fused ssim would be nice to add to the readme without changing much. Any special instructions would be appreciated. If you want to do it, please go ahead and I'll merge the PR.

Would it be possible for you to create another file similar to `train_image.py` which trains a 3D voxel grid using SSIM over some a 3D voxel grid (perhaps of openly available medical data). Feel free to skip this if it is a hassle.

Regards,
Rahul

**Comment by PaPieta** [comment]
Hi again, 

In that case, I'll try to do the readme update somewhere within this or next week.

The ```train_image.py``` part shouldn't be a problem as well. But the resolution of the 3D image will probably be quite small to work well as a part of the repo.

Best,
Pawel
